pipelines/controller/src/main/java/org/openmrs/analytics/DWHFilesManager.java

```diff
@@ -0,0 +1,353 @@
+  }
+
+  private void setLastPurgeRunStatus(PurgeRunStatus status) {
+    if (status == PurgeRunStatus.SUCCESS) {
+      lastPurgeRunEnd = LocalDateTime.now();
+    }
+  }
+
+  private boolean isPurgeJobRunning() {
+    return isPurgeJobRunning;
+  }
+
+  private void setPurgeJobRunning(boolean isPurgeJobRunning) {
+    this.isPurgeJobRunning = isPurgeJobRunning;
+  }
+
+  private synchronized boolean lockPurgeJob() {
+    if (isPurgeJobRunning()) {
+      // Already the job is running so lock can't be acquired.
+      return false;
+    }
+    setPurgeJobRunning(true);
+    return true;
+  }
+
+  private synchronized void releasePurgeJob() {
+    setPurgeJobRunning(false);
+  }
+
+  /**
+   * @return the next scheduled time to run the purge job based on the previous run time.
+   */
+  LocalDateTime getNextPurgeTime() {
+    if (lastPurgeRunEnd == null) {
+      return LocalDateTime.now();
+    }
+    return purgeCron.next(lastPurgeRunEnd);
+  }
+
+  // Every 30 seconds, check if the purge job needs to be triggered.
+  @Scheduled(fixedDelay = 30000)
+  private void checkPurgeSchedule() {
+    try {
+      if (!lockPurgeJob()) {
+        return;
+      }
+      LocalDateTime next = getNextPurgeTime();
+      logger.info("Last purge run was at {} next run is at {}", lastPurgeRunEnd, next);
+      if (next.compareTo(LocalDateTime.now()) <= 0) {
+        logger.info("Purge run triggered at {}", LocalDateTime.now());
+        purgeDWHFiles();
+        logger.info("Purge run completed at {}", LocalDateTime.now());
+      }
+    } finally {
+      releasePurgeJob();
+    }
+  }
+
+  /**
+   * This method will scan the DWH base directory to check for any older DWH snapshots that needs to
+   * be purged. It will retain the most recent successfully completed snapshots and purge the older
+   * ones. Any incomplete snapshots will not be purged by this job and will have to be manually
+   * verified and acted upon.
+   */
+  private void purgeDWHFiles() {
+    String rootPrefix = dataProperties.getDwhRootPrefix();
+    Preconditions.checkState(rootPrefix != null && !rootPrefix.isEmpty());
+
+    int lastNSnapshotsToRetain = dataProperties.getLastNDWHSnapshotsToRetain();
+    Preconditions.checkState(
+        lastNSnapshotsToRetain > 0, "lastNSnapshotsToRetain should be positive value");
+    String baseDir = getBaseDir(rootPrefix);
+    try {
+      String prefix = getPrefix(rootPrefix);
+      List<ResourceId> paths =
+          getAllChildDirectories(baseDir).stream()
+              .filter(dir -> dir.getFilename().startsWith(prefix))
+              .collect(Collectors.toList());
+
+      TreeSet<String> recentSnapshotsToBeRetained =
+          getRecentSnapshots(paths, lastNSnapshotsToRetain);
+      deleteOlderSnapshots(paths, recentSnapshotsToBeRetained);
+      setLastPurgeRunStatus(PurgeRunStatus.SUCCESS);
+    } catch (IOException e) {
+      setLastPurgeRunStatus(PurgeRunStatus.FAILURE);
```

Comment:
Does this call have any affect?

Reply:
Yes, this has no effect. I have refactored the code now to not have these lines.